core/src/main/java/com/linecorp/armeria/common/CompositeHttpHeadersBase.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.linecorp.armeria.common.StringMultimap.DEFAULT_SIZE_HINT;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.AsciiString;
+
+/**
+ * The composite container implementation which wraps {@link HttpHeadersBase}s
+ * to avoid expensive copy operations.
+ */
+class CompositeHttpHeadersBase
+        extends CompositeStringMultimap</* IN_NAME */ CharSequence, /* NAME */ AsciiString>
+        implements HttpHeaderGetters {
+
+    private static final Supplier<StringMultimap<CharSequence, AsciiString>> DEFAULT_APPENDER_SUPPLIER =
+            () -> new HttpHeadersBase(DEFAULT_SIZE_HINT);
+
+    CompositeHttpHeadersBase(HttpHeaderGetters... parents) {
+        super(from(parents), DEFAULT_APPENDER_SUPPLIER);
+    }
+
+    CompositeHttpHeadersBase(@Nullable List<HttpHeaderGetters> additionals,
```

Comment:
Considering the current usage, which involves three different headers in HttpHeadersUtil, it seems that this constructor isn't necessary. Do you have a specific reason for adding it?

Reply:
> in HttpHeadersUtil, it seems that this constructor isn't necessary.

https://github.com/line/armeria/blob/c6b89064dfadfb985845876da86692e01e8bf1ec/core/src/main/java/com/linecorp/armeria/internal/common/HttpHeadersUtil.java#L57-L62

If we don't have this constructor, how can we pass `additioanls` headers to `CompositeStringMultimap`? 🤔

https://github.com/line/armeria/blob/c6b89064dfadfb985845876da86692e01e8bf1ec/core/src/main/java/com/linecorp/armeria/common/AbstractHttpHeadersBuilder.java#L32-L40

https://github.com/line/armeria/blob/c6b89064dfadfb985845876da86692e01e8bf1ec/core/src/main/java/com/linecorp/armeria/common/DefaultHttpHeadersBuilder.java#L24-L30

Also, this constructor is needed for next PR, to add `CompositeHttpHeaderBuilder` similar with `HttpHeaderBuilder`!
It uses `new HttpHeaderBase(..)` constuructor.

Reply:
> If we don't have this constructor, how can we pass additioanls headers to CompositeStringMultimap?

```java
CompositeHttpHeadersBase(HttpHeaderGetters... parents) {...}
```
This constructor accepts multiple `HttpHeaders`. Can't users simply specify the additional headers that take precedence as the first argument?
```java
/**
 * Merges the given {@link ResponseHeaders}. The headers have priority in the following order.
 * <pre>{@code
 * additional headers (highest priority) > headers > default headers > framework headers (lowest priority)
 * }</pre>
 */
public static ResponseHeaders mergeResponseHeaders(ResponseHeaders headers,
                                                   HttpHeaders additionalHeaders,
                                                   HttpHeaders defaultHeaders) {
    ...
    new CompositeHttpHeadersBase(additionalHeaders, headers, defaultHeaders);
    ...
}
```

> this constructor is needed for next PR, to add CompositeHttpHeaderBuilder similar with HttpHeaderBuilder!
It uses new HttpHeaderBase(..) constructor

I might miss this. Could you share an example of how it will be used?